### PR TITLE
Release 0.19.3

### PR DESCRIPTION
**Pull request by TobiGr**
# Changelog
### New
- Search on YouTube Music #3240
- Basic Android TV support #2806 

### Improved
- Improved checking for new version #3344
- Avoid unnecessary changes to the upload date of saved streams #3370
- Save and restore playback parameters into/from preferences #3288
- Show message when content isn't supported yet instead of crashing #3300
- Improved size handling of the drawer header title #3412
- Improved popup player resize with pinch gestures #3075
- Enqueue streams on long press on background and popup buttons in channel #3456 
- Added the ability to remove all watched videos from a local playlist #3065

### Fixed
- Fixed visibility of group sort button in the subscriptions fragment #3374
- Fixed detection of network related exceptions #3294
- Fixed age restricted content setting not working #3410
- Fixed certain kinds of reCAPTCHAs #3414
- Fixed crash when opening bookmarks while playlist is `null` #3345
- Fixed escaping in JSON created by the crash reporter by using nanojson instead of org.json #3463

### Development
- Added Checkstyle to build scripts & code style improvements #3282, 3464
- Fixed F-Droid build by ensuring the debug apk file name is used only in debug builds #3415
- Force UTF-8 encoding for Gradle #3460

### New
- Added support for YouTube Music search TeamNewPipe/NewPipeExtractor#291
- Throw the new `ContentNotSupportedException` when the content is not yet supported TeamNewPipe/NewPipeExtractor#300
- Added support for PeerTube's video channels TeamNewPipe/NewPipeExtractor#308

### Improved
- [SoundCloud] Added supported countries TeamNewPipe/NewPipeExtractor#299
- Use our [nanojson fork](https://github.com/TeamNewPipe/nanojson) to support JS-like JSON TeamNewPipe/NewPipeExtractor#317 

### Fixed
- Use media.ccc.de instead of api.media.ccc.de TeamNewPipe/NewPipeExtractor#301

--------
## ToDo

- [x] fix rotation regression
- [x] update nanojson or use our fork

--------

# APK for testing
[NewPipe_v0.19.3_RC4.apk.zip](https://github.com/TeamNewPipe/NewPipe/files/4570250/NewPipe_v0.19.3_RC4.apk.zip)

Please report new bugs in the corresponding issue #3479

### CODE REVIEW

File: fastlane/metadata/android/en-US/changelogs/930.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+• Basic Android TV support
+
+Improved
+• Added the ability to remove all watched videos from a local playlist
+• Show message when content isn't supported yet instead of crashing
+• Improved popup player resize with pinch gestures
+• Enqueue streams on long press on background and popup buttons in channel
+• Improved size handling of the drawer header title
+
+### Fixed
```
Comment (wb9688) [inline fastlane/metadata/android/en-US/changelogs/930.txt:12]: Why does this have ### before it and the others not? Also, why do we use bullet points in some changelogs while using hyphens in others?

File: fastlane/metadata/android/en-US/changelogs/930.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+• Fixed age restricted content setting not working
+• Fixed certain kinds of reCAPTCHAs
+• Fixed crash when opening bookmarks while playlist is `null`
+• Fixed detection of network related exceptions
+• Fixed visibility of group sort button in the subscriptions fragment
+
+and more
```
Comment (wb9688) [inline fastlane/metadata/android/en-US/changelogs/930.txt:19]: Why no newline at end of file?

File: fastlane/metadata/android/en-US/changelogs/930.txt (hunk 1 above)
Comment (TobiGr) [reply fastlane/metadata/android/en-US/changelogs/930.txt:12]: Thanks, I forgot to remove the ###.
Well, F-Droid's website supports markdown, while the app does not. I didn't knew this when I wrote the first chagelogs. That's why I switched to UTF-8 characters and used .txt files instead.

**Comment by ondernuray** [comment]
## Exception
* __User Action:__ requested stream
* __Request:__ https://www.youtube.com/watch?v=P3LEcaLWor0
* __Content Language:__ TR
* __Service:__ YouTube
* __Version:__ 0.19.3
* __OS:__ Linux Android 10 - 29


<details><summary><b>Crash log</b></summary><p>

```
java.lang.NullPointerException: Attempt to invoke virtual method 'com.grack.nanojson.JsonObject com.grack.nanojson.JsonObject.getObject(java.lang.String)' on a null object reference
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.getErrorMessage(YoutubeStreamExtractor.java:580)
	at org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:81)
	at org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:62)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$getStreamInfo$3(ExtractorHelper.java:116)
	at org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$5fJcha6Sq5APJBLdG6osaJby-mc.call(Unknown Source:4)
	at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)
	at io.reactivex.Maybe.subscribe(Maybe.java:4154)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)
	at io.reactivex.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)
	at io.reactivex.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)
	at io.reactivex.Flowable.subscribe(Flowable.java:14479)
	at io.reactivex.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)
	at io.reactivex.Maybe.subscribe(Maybe.java:4154)
	at io.reactivex.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)
	at io.reactivex.Single.subscribe(Single.java:3438)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:919)

```
</p></details>
<hr>

**Comment by ondernuray** [comment]
After a live stream closed, i cannot watch offline record. you can see log files above.

**Comment by wb9688** [comment]
@ondernuray: That's not a regression, i.e. irrelevant in this thread. We don't support watching a livestream on YouTube for x amount of time that it takes YouTube to make it a normal stream. That issue depends on TeamNewPipe/NewPipeExtractor#273. Also please don't double-post like that next time.

**Pull request merged**